keps/sig-node/xxx-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1372 @@
+  Kubernetes and does not have to be used by plugins, therefore it is
+  not described further in this KEP.
+- Kubelet must be extended to retrieve information from ResourceClaims
+  and then invoke local plugin methods. It must pass information about
+  the additional resources to the container runtime. It must detect
+  whether the container runtime has the necessary support and
+  advertise that in the cluster via node labels to simplify deployment
+  of plugins and Pod scheduling.
+
+```
+<<[UNRESOLVED @pohly]>>
+All of the changes in Kubernetes need to be specified in a lot more detail.
+
+A flow diagram showing the state transitions of a ResourceClaim needs to be added.
+
+Upgrade and downgrade scenarios should already be considered for v1alpha1 to ensure
+that whatever changes will be needed are in place before going to v1beta1 where
+downgrades have to be supported.
+
+All of that will be added once there is consensus to move ahead with this proposal.
+<<[/UNRESOLVED]>>
+```
+
+The gRPC interface is inspired by
+[CSI](https://github.com/container-storage-interface/spec/blob/master/spec.md),
+with “volume” replaced by “resource” and volume specific parts
+removed.
+
+```
+<<[UNRESOLVED @pohly]>>
+Do plugin operations need secrets? They are currently not part of the proposed Kubernetes API.
+<<[/UNRESOLVED]>>
+```
+
+### Node Service RPC
+
+#### `NodePrepareResource`
+
+This RPC is called by kubelet when a Pod that wants to use the
+specified resource is scheduled on a node.  The Plugin SHALL assume
+that this RPC will be executed on the node where the resource will be
+used.  The Plugin SHALL return device name and kind for allocated
+device[s].
+
+The Plugin SHALL create or update json file[s] in CDI format for each
+allocated device. These files SHALL be used by runtime to update
+runtime configuration before creating containers that use the
+device[s].
+
+This operation SHALL do as little work as possible as it’s called
+after a pod is scheduled to a node. All potentially failing operations
+SHALL be done during allocation phase.
+
+This operation MUST be idempotent. If the resource corresponding to
+the `resource_id` has already been prepared, the Plugin MUST reply `0
+OK`.
+
+If this RPC failed, or kubelet does not know if it failed or not, it
+MAY choose to call `NodePrepareResource` again, or choose to call
+`NodeUnprepareResource`.
+
+After a successful call, Kubelet MUST pass device names and kinds to
+the runtime through the CRI protocol.
+
+```
+<<[UNRESOLVED @bart0sh]>>
+CRI protocol may need to be extended for
+this purpose, e.g. device id can be added to the CRI Device structure.
+<<[/UNRESOLVED]>>
+```
+
+```protobuf
+message NodePrepareResourceRequest {
+  // The UID of the ResourceClaim. This field is REQUIRED.
+  string resource_uid = 1;
+}
+
+message NodePrepareResourceResponse {
+  // The map <device name>:<kind> for all allocated devices
+ // where device name is a “name of the device” and kind is a
```

Comment:
@bart0sh, @kad: can you clarify for me what the "name of the device" is?

Is that some arbitrary string made up by the node plugin? How does it relate to the CSI JSON file?

Reply:
Why is this a map?

Reply:
Is the name required to be unique across all JSON files?

For example, suppose I have two JSON files:
```
{ "kind": "foo", "devices": [ { "name": "something" } ] }
```

```
{ "kind": "bar", "devices": [ { "name": "something" } ] }
```

Then with the current API, it is not possible to select both devices at the same.

Reply:
Resolved via https://github.com/pohly/enhancements/pull/5